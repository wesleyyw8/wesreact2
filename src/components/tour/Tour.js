import React, {PropTypes} from 'react';
import countDown from './countDown';

class Tour extends React.Component {
  constructor(props, context) {
    super(props, context);
    this.state = {
      // timer: Object.assign({}, this.props.tour.timer)
      timer: this.props.tour.timer
    };
  }
  componentDidMount() {
    const { groupIndex, tourIndex } = this.props;
    const x = setInterval(() => {
      let timer = countDown(this.state.timer);
      if (timer.getTime() === new Date().getTime()) {
        this.props.onCountDownFinished(groupIndex, tourIndex);
        clearInterval(x);
      }
      this.setState({timer});
    }, 1000);

  }
  formatTime(time) {
    const minutes = time.getMinutes();
    const seconds = time.getSeconds();
    return {
      years: time.getFullYear() - new Date().getFullYear(),
      months: time.getMonth() - new Date().getMonth(),
      days: time.getDate(),
      hours: time.getHours(),
      minutes: minutes < 10 ? `0${minutes}` : minutes,
      seconds: seconds  < 10 ? `0${seconds}` : seconds
    };
  }
  render() {
    const { tourName, seatsLeft, isAvailable } = this.props.tour;
    const { groupIndex, tourIndex } = this.props;
    const { onBookingNowClick } = this.props;

    const timer = this.formatTime(this.state.timer);
    const onBookingButton = 
      (<button 
        onClick={() => onBookingNowClick(groupIndex, tourIndex, seatsLeft)}
        disabled={!isAvailable}
        className={'btn btn-primary'} >
          Book now!
      </button>);

    return (
      <div className="tour col-sm-3">
        <div className="card">
          <div className="card-header text-center">
            {`${timer.years > 0 ? `${timer.years} year ` : ''}`}
            {`${timer.months > 0 ? `${timer.months} months ` : ''}`}
            {`${timer.days > 0 ? `${timer.days} days and ` : ''}`} 
            {timer.hours}:{timer.minutes}:{timer.seconds}
          </div>
          <img className="card-img-top" src="https://cdn.tourradar.com/s3/tour/645x430/24087_1ac9705f.jpg" alt="Card image cap" />
          <div className="card-body">
            <h4 className="card-title">{tourName}</h4>
            <p className="card-text">Seats left: {seatsLeft}</p>
            {onBookingButton}
          </div>
        </div>
      </div>
    );
  }
}

Tour.propTypes = {
  tour: PropTypes.object.isRequired,
  onBookingNowClick: PropTypes.func.isRequired,
  groupIndex: PropTypes.number.isRequired,
  tourIndex: PropTypes.number.isRequired,
  onCountDownFinished: PropTypes.func.isRequired
};

export default Tour;