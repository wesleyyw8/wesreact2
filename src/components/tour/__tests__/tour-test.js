import React from 'react';
import {mount, shallow} from 'enzyme';
import Tour from '../Tour';

describe('Tour component', () => {
  const setup = () => {
    const props = {
      key: 0,
      tourIndex: 0,
      groupIndex: 0,
      tour: {
        tourName: 'test',
        seatsLeft: 2,
        isAvailable: true,
        timer: {
          days: 3,
          hours: 2,
          minutes: 3,
          seconds: 2
        }
      },
      onCountDownFinished: () => {},
      onBookingNowClick: () => {}
    };
    return shallow(
      <Tour {...props} />);
  };

  test('Card Title has only one item and its content should be the tourName', () => {
    const wrapper = setup();
    expect(wrapper.find('h4').length).toEqual(1);
    expect(wrapper.find('h4').text()).toEqual('test');
  });

  test('Card Title has only one item and its content should be the tourName', () => {
    const wrapper = setup();
    expect(wrapper.find('.card-text').length).toEqual(1);
    expect(wrapper.find('.card-text').text()).toEqual('Seats left: 2');
  });
  test('Card Header should have the date formated', () => {
    const wrapper = setup();
    expect(wrapper.find('.card-header.text-center').text()).toEqual('3 days 2:03:02');
  });
});