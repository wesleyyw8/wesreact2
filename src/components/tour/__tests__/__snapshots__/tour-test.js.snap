// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tour component Card Header should have the date formated 1`] = `
ShallowWrapper {
  "length": 1,
  "node": <div
    className="tour col-sm-3"
  >
    <div
      className="card"
    >
      <div
        className="card-header text-center"
      >
        
        00
        :
        00
        :
        00
      </div>
      <img
        alt="Card image cap"
        className="card-img-top"
        src="https://cdn.tourradar.com/s3/tour/645x430/24087_1ac9705f.jpg"
      />
      <div
        className="card-body"
      >
        <h4
          className="card-title"
        >
          test
        </h4>
        <p
          className="card-text"
        >
          Seats left: 
          2
        </p>
        <button
          className="btn btn-primary"
          disabled={false}
          onClick={[Function]}
        >
          Book now!
        </button>
      </div>
    </div>
  </div>,
  "nodes": Array [
    <div
      className="tour col-sm-3"
    >
      <div
        className="card"
      >
        <div
          className="card-header text-center"
        >
          
          00
          :
          00
          :
          00
        </div>
        <img
          alt="Card image cap"
          className="card-img-top"
          src="https://cdn.tourradar.com/s3/tour/645x430/24087_1ac9705f.jpg"
        />
        <div
          className="card-body"
        >
          <h4
            className="card-title"
          >
            test
          </h4>
          <p
            className="card-text"
          >
            Seats left: 
            2
          </p>
          <button
            className="btn btn-primary"
            disabled={false}
            onClick={[Function]}
          >
            Book now!
          </button>
        </div>
      </div>
    </div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_context": Object {},
      "_currentElement": <Tour
        groupIndex={0}
        onBookingNowClick={[Function]}
        onCountDownFinished={[Function]}
        tour={
          Object {
            "isAvailable": true,
            "seatsLeft": 2,
            "timer": 2020-03-01T00:00:00.000Z,
            "tourName": "test",
          }
        }
        tourIndex={0}
      />,
      "_instance": Tour {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "groupIndex": 0,
          "onBookingNowClick": [Function],
          "onCountDownFinished": [Function],
          "tour": Object {
            "isAvailable": true,
            "seatsLeft": 2,
            "timer": 2020-03-01T00:00:00.000Z,
            "tourName": "test",
          },
          "tourIndex": 0,
        },
        "refs": Object {},
        "state": Object {
          "distance": 0,
          "timer": 2020-03-01T00:00:00.000Z,
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 3,
      "_nativeContainerInfo": null,
      "_nativeParent": null,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": Object {
        "_currentElement": <div
          className="tour col-sm-3"
        >
          <div
            className="card"
          >
            <div
              className="card-header text-center"
            >
              
              00
              :
              00
              :
              00
            </div>
            <img
              alt="Card image cap"
              className="card-img-top"
              src="https://cdn.tourradar.com/s3/tour/645x430/24087_1ac9705f.jpg"
            />
            <div
              className="card-body"
            >
              <h4
                className="card-title"
              >
                test
              </h4>
              <p
                className="card-text"
              >
                Seats left: 
                2
              </p>
              <button
                className="btn btn-primary"
                disabled={false}
                onClick={[Function]}
              >
                Book now!
              </button>
            </div>
          </div>
        </div>,
        "_renderedOutput": <div
          className="tour col-sm-3"
        >
          <div
            className="card"
          >
            <div
              className="card-header text-center"
            >
              
              00
              :
              00
              :
              00
            </div>
            <img
              alt="Card image cap"
              className="card-img-top"
              src="https://cdn.tourradar.com/s3/tour/645x430/24087_1ac9705f.jpg"
            />
            <div
              className="card-body"
            >
              <h4
                className="card-title"
              >
                test
              </h4>
              <p
                className="card-text"
              >
                Seats left: 
                2
              </p>
              <button
                className="btn btn-primary"
                disabled={false}
                onClick={[Function]}
              >
                Book now!
              </button>
            </div>
          </div>
        </div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": null,
      "_topLevelWrapper": null,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Tour
    groupIndex={0}
    onBookingNowClick={[Function]}
    onCountDownFinished={[Function]}
    tour={
      Object {
        "isAvailable": true,
        "seatsLeft": 2,
        "timer": 2020-03-01T00:00:00.000Z,
        "tourName": "test",
      }
    }
    tourIndex={0}
  />,
}
`;
